recipes/opencv/4.x/conanfile.py

```diff
@@ -82,7 +82,11 @@ def requirements(self):
         elif self.options.with_jpeg == "libjpeg-turbo":
             self.requires("libjpeg-turbo/2.0.5")
         if self.options.with_jpeg2000 == "jasper":
-            self.requires("jasper/2.0.21")
+            if tools.Version(self.version) >= "4.5.0":
+                self.requires("jasper/2.0.21")
+            else:
+                #https://github.com/opencv/opencv/issues/17984
+                self.requires("jasper/2.0.16")
```

Comment:
It should raise in `build()` if this condition is not honored (overridden).

@uilianries @SSE4 @danimtb @jgsogo if we could indicate version range of each dependency in conandata.yml, and loop over them in `build()` with an helper function, it would be easier and we could document compatible versions.

Reply:
Yes, this recipe needs to raise in the `validate` method after the graph has been resolved. The requirement to `jasper` can be overridden somewhere else (consumer).

I don't fully get the suggestion (that loop...). We should be able to open automatically PRs trying to upgrade dependencies (the objective would be to have a consistent graph, all packages use the same versions of the dependencies) and, of course, if this is done we need a place to write down versions that are incompatible so the bot doesn't open them over and over again.

Reply:
done

Reply:
build or validate? Its in build now, shall I move it?

Reply:
[`validate` is a better place for it](https://docs.conan.io/en/latest/reference/conanfile/methods.html?highlight=validate#validate). Without building the package Conan knows that the binary with the resulting configuration cannot be built. Of course, Conan will raise if you try to build the package, but `conan info` will show valuable information too: 

```
conan info ...
...
  ID: INVALID
```

Reply:
I thought that deps versions were not yet available at `validate()` time (only final deps options values for the moment)?

Reply:
moved to validate 🤷

Reply:
bad idea, that does not even work, seems that `deps_cpp_info` is not available in `validate`:
```
ERROR: opencv/4.1.2: Error in validate() method, line 275
        if tools.Version(self.version) < "4.5.0" and self.deps_cpp_info["jasper"].version > "2.0.16":
        KeyError: 'jasper'
```

Reply:
It should go to `build()` for the moment.